File: app/src/main/java/im/tox/antox/ToxDataFile.java
package im.tox.antox;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

import android.content.Context;
import android.os.Environment;

/**
 * Class for loading and saving the data file to be used by JTox(byte[] data,
 * ...)
 * 
 * @author Mark Winter (astonex)
 */
public class ToxDataFile {

	/**
	 * Other clients use the file name data so we shall as well
	 */
	private String fileName = "data";

	/* Some variables for debgging */
	private byte[] dataContent;
	private String dataContentString;

	public ToxDataFile() {
	}

	/**
	 * Method to check if the data file exists before attempting to use it
	 * 
	 * @param ctx
	 * @return
	 */
	public boolean doesFileExist() {
		File myFile = new File("/sdcard/" + fileName);
		return myFile.exists();
		  //Do action
		
//		File file = ctx.getFileStreamPath(fileName);
//		return file.exists();
	}

	/**
	 * Method for deleting the tox data file
	 * 
	 * @param ctx
	 */
	public void deleteFile() {
		File file = new File("/sdcard/" + fileName);
		file.delete();
	}

	/**
	 * Method for loading data from a saved file and return it. Requires the
	 * context of the activity or service calling it.
	 * 
	 * @param ctx
	 * @return
	 */
	public byte[] loadFile() {
		FileInputStream fin = null;
		final File file = new File(Environment.getExternalStorageDirectory()
				.getAbsolutePath(), fileName);
		byte[] data = null;
		try {
			fin = new FileInputStream(file);
			data = new byte[(int) file.length()];
			fin.read(data);

		} catch (FileNotFoundException e) {
			e.printStackTrace();

		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			try {
				if (fin != null) {
					fin.close();
				}
			} catch (IOException ioe) {
				System.out.println("Error while closing stream: " + ioe);
			}
		}
		return data;
		// File dataFile = ctx.getFileStreamPath(fileName);
		// int sizeOfFile = (int) dataFile.length();
		// byte[] data = new byte[sizeOfFile];
		// try {
		// BufferedInputStream buff = new BufferedInputStream(new
		// FileInputStream(dataFile));
		// buff.read(data, 0, data.length);
		// buff.close();
		// } catch (IOException e) {
		// e.printStackTrace();
		// }
		//
		// return data;

	}

	/**
	 * Method for saving tox data. Requires the data to be saved and the context
	 * of the activity or service calling it
	 * 
	 * @param dataToBeSaved
	 * @param ctx
	 */
	public void saveFile(byte[] dataToBeSaved) {
		File myFile = new File("/sdcard/" + fileName);
		try {
			myFile.createNewFile();
		} catch (IOException e1) {
			// TODO Auto-generated catch block
			e1.printStackTrace();
		}
		try {
			FileOutputStream output = new FileOutputStream(myFile);
			output.write(dataToBeSaved, 0, dataToBeSaved.length);
			output.close();

		} catch (IOException e) {
			e.printStackTrace();
		}
		// try {
		// outputStream = ctx.openFileOutput(fileName, Context.MODE_PRIVATE);
		// outputStream.write(dataToBeSaved);
		// outputStream.close();
		// } catch (IOException e) {
		// e.printStackTrace();
		// }
	}
}

File: app/src/main/scala/chat/tox/antox/tox/ToxDataFile.scala
package chat.tox.antox.tox

import java.io._

import android.content.Context
import chat.tox.antox.utils.{AntoxLog, FileUtils}
import im.tox.tox4j.core.options.SaveDataOptions
import im.tox.tox4j.core.options.SaveDataOptions.ToxSave
import im.tox.tox4j.impl.jni.ToxCryptoImpl
import org.scaloid.common.LoggerTag

object ToxDataFile {

  def isEncrypted(data: Array[Byte]): Boolean = {
    ToxCryptoImpl.isDataEncrypted(data)
  }

  def isEncrypted(file: File): Boolean = {
    var fin: FileInputStream = null
    var data: Array[Byte] = null
    try {
      fin = new FileInputStream(file)
      data = Array.ofDim[Byte](file.length.toInt)
      fin.read(data)
    } catch {
      case e: FileNotFoundException => e.printStackTrace()
      case e: IOException => e.printStackTrace()
    } finally {
      try {
        if (fin != null) {
          fin.close()
        }
      } catch {
        case ioe: IOException => ioe.printStackTrace()
      }
    }
    ToxDataFile.isEncrypted(data)
  }


}

class ToxDataFile(ctx: Context, fileName: String) {

  private val TAG = LoggerTag(getClass.getName)

  def isEncrypted: Boolean = {
    ToxDataFile.isEncrypted(ctx.getFileStreamPath(fileName))
  }

  def decrypt(pass: String): Unit = {
    if (isEncrypted) {
      val data = loadFile()
      val salt = ToxCryptoImpl.getSalt(data)
      saveFile(ToxCryptoImpl.decrypt(data, ToxCryptoImpl.passKeyDeriveWithSalt(pass.getBytes, salt)))
    }
  }

  def doesFileExist(): Boolean = {
    AntoxLog.debug("fileName: " + fileName, TAG)
    val selfFile = ctx.getFileStreamPath(fileName)
    if (selfFile == null) {
      AntoxLog.debug("selfFile is null!", TAG)
    }
    selfFile.exists()
  }

  def exportFile(dest: File): Unit = {
    if (!dest.exists()) {
      throw new IllegalArgumentException("dest must exist")
    }

    FileUtils.copy(ctx.getFileStreamPath(fileName), new File(dest + "/" + fileName + ".tox"))
  }

  def deleteFile() {
    ctx.deleteFile(fileName)
  }

  def loadFile(): Array[Byte] = {
    var fin: FileInputStream = null
    val file = ctx.getFileStreamPath(fileName)
    var data: Array[Byte] = null
    try {
      fin = new FileInputStream(file)
      data = Array.ofDim[Byte](file.length.toInt)
      fin.read(data)
    } catch {
      case e: FileNotFoundException => e.printStackTrace()
      case e: IOException => e.printStackTrace()
    } finally {
      try {
        if (fin != null) {
          fin.close()

        }
      } catch {
        case ioe: IOException => ioe.printStackTrace()
      }
    }
    data
  }

  def loadEncryptedFile(pass: String): Array[Byte] = {
    var fin: FileInputStream = null
    val file = ctx.getFileStreamPath(fileName)
    var data: Array[Byte] = null
    try {
      fin = new FileInputStream(file)
      data = Array.ofDim[Byte](file.length.toInt)
      fin.read(data)
    } catch {
      case e: FileNotFoundException => e.printStackTrace()
      case e: IOException => e.printStackTrace()
    } finally {
      try {
        if (fin != null) {
          fin.close()
        }
      } catch {
        case ioe: IOException => ioe.printStackTrace()
      }
    }
    val salt = ToxCryptoImpl.getSalt(data)
    ToxCryptoImpl.decrypt(data, ToxCryptoImpl.passKeyDeriveWithSalt(pass.getBytes, salt))
  }

  def loadAsSaveType(): SaveDataOptions = {
    if (doesFileExist()) {
      ToxSave(loadFile())
    } else {
      SaveDataOptions.None
    }
  }

  def saveFile(dataToBeSaved: Array[Byte]) {
    val myFile = ctx.getFileStreamPath(fileName)
    try {
      myFile.createNewFile()
    } catch {
      case e1: IOException => e1.printStackTrace()
    }
    try {
      val output = new FileOutputStream(myFile)
      output.write(dataToBeSaved, 0, dataToBeSaved.length)
      output.close()
    } catch {
      case e: IOException => e.printStackTrace()
    }
  }
}
